Add fuzzer for RowNumber operator

duanmeng opened the pull request:
The RowNumberFuzzer is a testing tool that automatically generates equivalent
query plans and then execute these plans to validate the consistency of the results.

It works as follows:

1. Data Generation: It starts by generating a random set of input data, also known as a vector.
   This data can have a variety of encodings and data layouts to ensure thorough testing.
2. Plan Generation: Generate two equivalent query plans, one is row-number over ValuesNode
    and the other is over TableScanNode.
3. Query Execution: Executes those equivalent query plans using the generated data and asserts
    that the results are consistent across different plans.
4. Iteration: This process is repeated multiple times to ensure reliability and robustness.

xiaoxmeng commented on velox/exec/tests/RowNumberFuzzer.cpp:
```diff
@@ -0,0 +1,541 @@
+  return i >= FLAGS_steps;
+}
+
+std::vector<RowVectorPtr> flatten(const std::vector<RowVectorPtr>& vectors) {
```
```
const std::vector<RowVectorPtr>&
```

xiaoxmeng commented on velox/exec/tests/RowNumberFuzzer.cpp:
```diff
@@ -0,0 +1,541 @@
+  std::vector<std::string> names;
+  std::vector<TypePtr> types;
+  for (auto i = 0; i < numKeys; ++i) {
+    names.push_back(fmt::format("c{}", i));
```
```
nit: /*maxDepth=*/1
```

xiaoxmeng commented on velox/exec/tests/RowNumberFuzzer.cpp:
```diff
@@ -0,0 +1,541 @@
+    const std::vector<std::string>& partitionKeys,
+    const std::vector<RowVectorPtr>& input) {
+  auto planNodeIdGenerator = std::make_shared<core::PlanNodeIdGenerator>();
+  std::vector<std::string> projectFields = partitionKeys;
```
Shall we also test the case without partition keys? Thanks!

xiaoxmeng commented on velox/exec/tests/RowNumberFuzzer.cpp:
```diff
@@ -0,0 +1,541 @@
+                  .rowNumber(partitionKeys)
+                  .project(projectFields)
+                  .planNode();
+  return PlanWithSplits{plan};
```
```
PlanWithSplits{std::move(plan)};
```

xiaoxmeng commented on velox/exec/tests/RowNumberFuzzer.cpp:
```diff
@@ -0,0 +1,541 @@
+
+void RowNumberFuzzer::verify() {
+  auto [keyNames, keyTypes] = generatePartitionKeys();
+  auto input = generateInput(keyNames, keyTypes);
```
const input and flatInput? thanks!

xiaoxmeng commented on velox/exec/tests/RowNumberFuzzer.cpp:
```diff
@@ -0,0 +1,541 @@
+  }
+
+  std::vector<PlanWithSplits> altPlans;
+  altPlans.push_back(makeDefaultPlan(keyNames, input));
```
Can we just use defaultPlan here instead of creating a new one?

xiaoxmeng commented on velox/exec/tests/RowNumberFuzzer.cpp:
```diff
@@ -0,0 +1,541 @@
+  const auto expected = execute(defaultPlan, /*injectSpill=*/false);
+
+  // If OOM injection is not enabled verify the results against DuckDB.
+  if (!FLAGS_enable_oom_injection) {
```
Shall we just check if expected is not null here?

xiaoxmeng commented on velox/exec/tests/RowNumberFuzzer.cpp:
```diff
@@ -0,0 +1,541 @@
+  // Wait for the task to be destroyed before start next query execution to
+  // avoid the potential interference of the background activities across query
+  // executions.
+  waitForAllTasksToBeDeleted();
```
Shall we put waitForAllTasksToBeDeleted() in folly::makeGuard to call that in case of exception as well? Thanks!

xiaoxmeng reviewed (commented):
@duanmeng looks pretty good % minors. Thanks!

xiaoxmeng reviewed (approved):
@duanmeng LGTM % minors. Thanks!

mbasmanova commented on velox/exec/tests/RowNumberFuzzer.h:
```diff
@@ -0,0 +1,22 @@
+#include <cstddef>
+
+namespace facebook::velox::exec::test {
+void rowNumberFuzzer(size_t seed);
```
Would it make sense to move this to velox/exec/fuzzer ?

duanmeng replied:
Got it, done.

mbasmanova commented on velox/docs/develop/testing/rowNumberFuzzer.rst:
```diff
@@ -0,0 +1,53 @@
+  #. An alter plan that uses TableScanNode as the source.
+3. Query Execution: Executes those equivalent query plans using the generated data and asserts that the results are
+   consistent across different plans.
+  i. Execute the base plan, compare the result with the reference (DuckDB) and use it as the expected result.
```
Why DuckDB and not Presto?

duanmeng replied:
Fixed by adding support for both DuckDbQueryRunner and PrestoQueryRunner.

mbasmanova commented on velox/docs/develop/testing/rowNumberFuzzer.rst:
```diff
@@ -0,0 +1,53 @@
+================
```
Should we rename this file to be consistent with other files? row-number-fuzzer.rst?

duanmeng replied:
Fixed.

mbasmanova reviewed (commented):
Thanks.

CC: @kgpai @kagamiori

duanmeng commented:
@mbasmanova @xiaoxmeng  I've updated the PR description, resolved all the comments, and added supporting of RowNumber in both DuckDbQueryRunner and PrestoQueryRunner. Could you please help take another look? Thanks.

mbasmanova commented:
@duanmeng Thank you for updating PR description.

> Plan Generation: Generate a variety of logically equivalent query plans.

This is a bit generic statement and it creates an impression of the Fuzzer creating many plans, but in fact it only creates 2 plans: row-number over ValuesNode and over TableScanNode. It would be nice to make this clear both in PR description and documentation.

duanmeng commented:
> @duanmeng Thank you for updating PR description.
> 
> > Plan Generation: Generate a variety of logically equivalent query plans.
> 
> This is a bit generic statement and it creates an impression of the Fuzzer creating many plans, but in fact it only creates 2 plans: row-number over ValuesNode and over TableScanNode. It would be nice to make this clear both in PR description and documentation.

@mbasmanova I've updated the PR description and document according to your guidance.

mbasmanova commented:
@duanmeng Thanks.
